bftengine/src/bftengine/ClientsManager.hpp

```diff
@@ -218,9 +219,45 @@ class ClientsManager : public ResPagesClient<ClientsManager>, public IPendingReq
     bool committed = false;
   };
 
+  class RequestsInfo {
+   public:
+    void emplaceSafe(NodeIdType clientId, ReqId reqSeqNum, const std::string& cid);
+    bool removeRequestsOutOfBatchBoundsSafe(NodeIdType clientId, ReqId reqSequenceNum);
+    bool findSafe(ReqId reqSeqNum);
+    void clearSafe();
+    void removeOldPendingReqsSafe(NodeIdType clientId, ReqId reqSeqNum);
+    void removePendingForExecutionRequestSafe(NodeIdType clientId, ReqId reqSeqNum);
+
+    size_t size() const { return requestsMap_.size(); }
+    bool find(ReqId reqSeqNum) const;
+    bool isPending(ReqId reqSeqNum) const;
+    void markRequestAsCommitted(NodeIdType clientId, ReqId reqSeqNum);
+    void infoOfEarliestPendingRequest(Time& earliestTime, RequestInfo& earliestPendingReqInfo) const;
+    void logAllPendingRequestsExceedingThreshold(const int64_t threshold,
```

Comment:
If I understand correctly we use the non locking functions when we designate a given entry form **clientsInfo_** to a thread and we don't access it from others?

Reply:
Not sure I understand the question. Doesn't the "Problem overview" part answer your question?

Reply:
Yes it does, but because the actual usage is something I am not in-depth familiar with, wanted to make sure. If my general understanding is correct we need to have a mechanism that blocks other threads from modifying the map **requestsMap_** while we are operating on it with the non-sync alternatives of the methods?

Reply:
Correct